### DISCUSSION TITLE
Node graph Support

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#186] [General]
# Node Graph

You can generate Node graph panel data with `CSV`, `JSON` or `XML` feature of Infinity data source.

![image](https://user-images.githubusercontent.com/153843/139112671-dee6a9aa-5165-4526-bc8a-7ce36c45181e.png)

From v0.8.0, we have out of the box support for node graph. Prior to v0.8.0, you can use "Table" format to create node graph  with limited options.

Irrespective of the version, you need two queries to form a Node graph visualisation.

1. First query which lists the nodes. Should have mandatory `id` and `title` field. This also can have additional `arc__*` fields to specify the arcs value for the node.  
2. Second query which list the edges (relations). Should have three mandatory fields namely `id`, `source` and `target`

Read more about the node graph data structure [here](https://grafana.com/docs/grafana/latest/visualizations/node-graph/)

## Node Graph format

From v0.8.0, infinity plugin have dedicated support for node graphs. You need to create two different queries with the format of `nodes` and `edges` for each query.

### Nodes

Nodes query should have following fields.

* id*
* title
* subTitle
* arc__*
* detail__*

If you have any `arc__*` field, corresponding value (number) will be used as the nodes arc value. Note : Total value of row item should sum to 1.  If you need to specify a color for the `arc__*` field, add a new field called `arc__*_color`. 

### Edges

Edges query should have following fields.

* id*
* source*
* target*
* detail__*

works same as **Nodes**. But doesn't support `arc__*` fields.

## Example

Below is an example of nodes query

![image](https://user-images.githubusercontent.com/153843/139114480-bd5c8571-4374-4ec0-8af0-224cc73ba3d8.png)

Below is an example of edges query

![image](https://user-images.githubusercontent.com/153843/139114526-5ffabaf1-7722-4205-9347-9496a9485306.png)

Corresponding Node Graph is given below

![image](https://user-images.githubusercontent.com/153843/139114591-2203f878-bb7d-4111-aaac-fe3bbc99bd6b.png)

Corresponding CSV data is given below

Nodes : 

```csv
id,title,sub-title,cpu,memory,c_disk_size,d,c_disk_size color,d color,detail__hello
A,Server A,Application Server,12,10,0.1,0.9,blue,red,world
B,Server B,DB Server,90,87,0.1,0.9,blue,red,hello
C,Server C,Application Server,20,23,0.20,0.80,blue,red,hello
D,Server D,Middleware Server,47,98,0.90,0.10,blue,red,world
```

Edges :
```csv
id,source,target,mainStat,secondaryStat,detail__one
1,A,B,30,mb/s,abc
2,A,C,20,mb/s,def
3,B,D,24.2,mb/s,ghi
```

> Some of the fields of the CSV doesn't match the node graph format. So we used column alias to match the exact field. Also for CSV format, number formats needs to set explicity. Otherwise, they will be rendered as string.

**Comment by yosiasz** [comment]
Great stuff! But not working out for me with the colors. Here is my sample data

```
{
      "nodes" : [
        {
          "id": 1,
          "title": "nginx",
          "mainStat": 21233,
          "arc__passed": 0.283,
          "passed_color": "green",
          "arc__failed": 0.717,
          "failed_color": "red",
        },
        {
          "id": 2,
          "title": "serviceA",
          "mainStat": 12000,
          "arc__passed": 0.167,
          "passed_color": "green",
          "arc__failed": 0.833,
          "failed_color": "red"
        },
        {
          "id": 3,
          "title": "serviceB",
          "mainStat": 8233,
          "arc__passed": 0.486,
          "passed_color": "green",
          "arc__failed": 0.514,
          "failed_color": "red"
        }
      ],
      "edges" : {
        "id": "a",
        "source": 1,
        "target": 2
      }
    }
```

**Reply by yosiasz** [reply]
never mind, I got it to work!

**Reply by yesoreyeram** [reply]
Glad you found the solution. If you can, share the solution. others may get some some benefit. 

Btw, how hard it was to setup. I am trying to improve the node graph setup in future versions. Any feedback appreciated.

**Reply by yesoreyeram** [reply]
I assume the problem with your setup is about naming the color field. That have to be `arc__passed_color` instead `passed_color`. Am I right?

**Reply by yosiasz** [reply]
> Glad you found the solution. If you can, share the solution. others may get some some benefit.
> 
> Btw, how hard it was to setup. I am trying to improve the node graph setup in future versions. Any feedback appreciated.

it was not hard at all. Super easy.

**Reply by yosiasz** [reply]
> `arc__passed_color`

actually if in the data source the data is actually arc__passed_color it does not work. It has to be something not arc__passed_color that you then alias as arc__passed_color 

![image](https://user-images.githubusercontent.com/900951/170736640-e47ee256-3937-425b-aed4-71cfa34ba452.png)

**Reply by yosiasz** [reply]
How does infinity grab data from columns index wise or name wise?
column[0] picks column 1 from csv 
or does it do
column['id']

**Reply by yesoreyeram** [reply]
From the code I can see it is by name. https://github.com/yesoreyeram/grafana-infinity-datasource/blob/main/src/app/parsers/InfinityParser.ts#L68

**Reply by yosiasz** [reply]
sorry I mean to say reading any of the columns. So if I have following columns
name,age,address
and I use inline csv
but for the Add Columns part I use
address--->string
name--->string
age-->float
Will that work? :)

Thanks!
Waiting patiently to see Infinity work with a database query even a simple one? I can do a simple POC Feature PR if you want me to.

**Reply by yesoreyeram** [reply]
Yes. That will work.
I am not sure how the infinity can help in database world. Supporting different database may require different drivers. Also there is already sqlyze plugin which aims to do something similar in that world. 
I am also fine with refining the idea. But still thinking the feasibility

**Reply by yosiasz** [reply]
I agree. there are other plugins that fill this space. Thanks!

**Reply by yosiasz** [reply]
I need to understand this part to translate it to the use of mysql as a data source
```
if (matching_color_field) {
              field.config = {
                displayName: field.name.replace('arc__', ''),
                color: {
                  mode: FieldColorMode?.Fixed || 'fixed',
                  fixedColor: matching_color_field.values.get(0) || '',
                },
              };
```

I assume it has something to do with nodeGraph node parameters.
Any field prefixed with arc__ will be used to create the color circle around the node. All values in these fields should add up to 1. You can specify color using config.color.fixedColor

Could you please point me to some documentation (other than the above) what this config does?

**Reply by yesoreyeram** [reply]
For the color, it seems every datasource needs to handle it explicitly. In the data frame fields returned by the datasource, it need to set a field config. This was exclusively handled by infinity. Though there is no way to handle config from the UI. For infinity, color config is a workaround.

https://grafana.com/docs/grafana/latest/visualizations/node-graph/

**Comment by yosiasz** [comment]
if infinity has the following feature, it would be game over for all other grafana plugins.

Ability to point the plugin to another mysql,mssql, anydb query. Game over 😉

**Reply by mshiladityam** [reply]
hey can you please help me render the context menu in NodeGraph? When I click on nodes/edges, nothing happens :(((

**Comment by mshiladityam** [comment]
Hey, this is great work. I built a really cool nodeGraph using this plugin. I only have one small problem.

Where does the detail__* data go? When I click on nodes and edges, nothing happens. Is it not supposed to show in the context menu? No context menu appears for me when I click on nodes or edges. 

Is this a bug?

**Reply by yesoreyeram** [reply]
See if this example helps to understand https://grafana-infinity-datasource.herokuapp.com/d/node-graph/node-graph?orgId=1

**Reply by yesoreyeram** [reply]
![Screenshot_20220615-081807](https://user-images.githubusercontent.com/153843/173766916-e4dce3b0-1ff6-48e3-b461-b284bfbe9536.png)

**Reply by mshiladityam** [reply]
Thanks a lot for your reply!!!
If we see the official documentation: https://grafana.com/docs/grafana/latest/visualizations/node-graph/#nodes
the context menu appears in the Graph layout itself. But it is not the case in your plugin. How to fix that?

Also, this is definitely misbehaving. you have define__hello as the column name, and "hello" as the field value of server B. "world" is the field value of server A. But in your screenshot, we have hello as the value of define, which doesn't make sense.

It is misbehaving for me too: the title of the next node is the value of the previous node.

**Reply by mshiladityam** [reply]
Also, if it renders only in Grid-view, what is the use-case of detail__* for edges?

**Reply by mshiladityam** [reply]
please help me!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! I am stuck and my project is near production!!

**Reply by mshiladityam** [reply]
> See if this example helps to understand https://grafana-infinity-datasource.herokuapp.com/d/node-graph/node-graph?orgId=1

Hey,
I think I got what your bug is.
It always prints, the value of the first detail, as the value of the header of the detail.
Also, the details for edges are never rendered, as they are only available on grid mode.

**Also, I want support for printing multiple lines in a detail. Currently it doesn t support newline characters.**

**Reply by mshiladityam** [reply]
hey please reply, I am stuck and really need your help :"((

**Reply by yesoreyeram** [reply]
Sorry buddy. Busy for couple of weeks. Try grafana community.

**Reply by mshiladityam** [reply]
If instead of mainStat or secondaryStat, I want a more human-readable label, what should I do? Is it possible?
